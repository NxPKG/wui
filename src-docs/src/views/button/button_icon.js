import React from 'react';

import {
  WuiButtonIcon,
  WuiFlexGroup,
  WuiFlexItem,
} from '../../../../src/components';

const colors = [
  'primary',
  'text',
  'accent',
  'subdued',
  'success',
  'warning',
  'danger',
  'disabled',
];

export default () => (
  <WuiFlexGroup gutterSize="s" alignItems="center">
    {colors.map(color => (
      <WuiFlexItem key={color} grow={false}>
        <WuiButtonIcon
          color={color}
          onClick={() => window.alert('Button clicked')}
          iconType="arrowRight"
          aria-label="Next"
          disabled={color === 'disabled' ? true : false}
        />
      </WuiFlexItem>
    ))}
  </WuiFlexGroup>
);
