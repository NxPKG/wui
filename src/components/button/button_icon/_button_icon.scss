.wuiButtonIcon {
  @include wuiButton;

  border: none;
  background-color: transparent;
  box-shadow: none;
  height: auto;
  min-height: $wuiSizeL;
  min-width: $wuiSizeL;
  line-height: 0; // ensures the icon creates the line-height and so it's always vertically centered
  padding: $wuiSizeXS;
  border-radius: $wuiBorderRadius;

  & > svg {
    // prevents the element and its children from receiving any pointer events to fix not bubbling click event in Safari
    // https://stackoverflow.com/questions/24078524/svg-click-events-not-firing-bubbling-when-using-use-element
    pointer-events: none;
  }

  &:disabled {
    color: $wuiButtonColorDisabledText;
    pointer-events: none;

    .wuiButtonIcon__icon {
      pointer-events: auto;
      cursor: not-allowed;
    }

    &:hover,
    &:focus {
      background-color: transparent;
      text-decoration: none;
    }
  }
}

// Create button modifiers based upon the map.
@each $name, $color in $wuiButtonIconTypes {
  .wuiButtonIcon--#{$name} {
    color: $color;

    &:focus {
      background-color: transparentize($color, .9);
    }

    &:hover,
    &:focus {
      @if ($name == 'disabled') {
        cursor: not-allowed;
      }
    }
  }
}
