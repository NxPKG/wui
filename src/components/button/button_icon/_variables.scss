// Modifier naming and colors.
$wuiButtonIconTypes: (
  accent: $wuiColorAccentText,
  danger: $wuiColorDangerText,
  disabled: $wuiButtonColorDisabledText,
  ghost: $wuiColorGhost,
  primary: $wuiColorPrimaryText,
  subdued: $wuiTextSubduedColor,
  success: $wuiColorSuccessText,
  text: $wuiTextColor,
  warning: $wuiColorWarningText,
);
