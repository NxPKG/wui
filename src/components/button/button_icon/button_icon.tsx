/*
 * Copyright 2022 Wazuh Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * NOTICE: THIS FILE HAS BEEN MODIFIED BY WAZUH INC UNDER COMPLIANCE WITH THE APACHE 2.0 LICENSE FROM THE ORIGINAL WORK
 * OF THE COMPANY Elasticsearch B.V.
 *
 * THE FOLLOWING IS THE COPYRIGHT OF THE ORIGINAL DOCUMENT:
 *
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, {
  AnchorHTMLAttributes,
  ButtonHTMLAttributes,
  FunctionComponent,
  Ref,
} from 'react';
import classNames from 'classnames';

import { getSecureRelForTarget } from '../../../services';
import {
  CommonProps,
  ExclusiveUnion,
  PropsForAnchor,
  PropsForButton,
  keysOf,
} from '../../common';

import { IconType, IconSize, WuiIcon } from '../../icon';

import { ButtonSize } from '../button';
import { validateHref } from '../../../services/security/href_validator';

export type WuiButtonIconColor =
  | 'accent'
  | 'danger'
  | 'disabled'
  | 'ghost'
  | 'primary'
  | 'subdued'
  | 'success'
  | 'text'
  | 'warning';

export interface WuiButtonIconProps extends CommonProps {
  iconType?: IconType;
  color?: WuiButtonIconColor;
  'aria-label'?: string;
  'aria-labelledby'?: string;
  isDisabled?: boolean;
  size?: ButtonSize;
  iconSize?: IconSize;
}

type WuiButtonIconPropsForAnchor = {
  type?: string;
} & PropsForAnchor<
  WuiButtonIconProps,
  {
    buttonRef?: Ref<HTMLAnchorElement>;
  }
>;

export type WuiButtonIconPropsForButton = {
  type?: 'submit' | 'reset' | 'button';
} & PropsForButton<
  WuiButtonIconProps,
  {
    buttonRef?: Ref<HTMLButtonElement>;
  }
>;

type Props = ExclusiveUnion<
  WuiButtonIconPropsForAnchor,
  WuiButtonIconPropsForButton
>;

const colorToClassNameMap: { [color in WuiButtonIconColor]: string } = {
  accent: 'wuiButtonIcon--accent',
  danger: 'wuiButtonIcon--danger',
  disabled: 'wuiButtonIcon--disabled',
  ghost: 'wuiButtonIcon--ghost',
  primary: 'wuiButtonIcon--primary',
  subdued: 'wuiButtonIcon--subdued',
  success: 'wuiButtonIcon--success',
  text: 'wuiButtonIcon--text',
  warning: 'wuiButtonIcon--warning',
};

export const COLORS = keysOf(colorToClassNameMap);

export const WuiButtonIcon: FunctionComponent<Props> = ({
  className,
  iconType,
  iconSize = 'm',
  color = 'primary',
  isDisabled: _isDisabled,
  href,
  type = 'button',
  target,
  rel,
  buttonRef,
  ...rest
}) => {
  const isHrefValid = !href || validateHref(href);
  const isDisabled = _isDisabled || !isHrefValid;

  const ariaHidden = rest['aria-hidden'];
  const isAriaHidden = ariaHidden === 'true' || ariaHidden === true;

  if (!rest['aria-label'] && !rest['aria-labelledby'] && !isAriaHidden) {
    console.warn(
      `WuiButtonIcon requires aria-label or aria-labelledby to be specified because icon-only
      buttons are screen-reader-inaccessible without them.`
    );
  }
  const classes = classNames(
    'wuiButtonIcon',
    colorToClassNameMap[color],
    className
  );

  // Add an icon to the button if one exists.
  let buttonIcon;

  if (iconType) {
    buttonIcon = (
      <WuiIcon
        className="wuiButtonIcon__icon"
        type={iconType}
        size={iconSize}
        aria-hidden="true"
      />
    );
  }

  // <a> elements don't respect the `disabled` attribute. So if we're disabled, we'll just pretend
  // this is a button and piggyback off its disabled styles.
  if (href && !isDisabled) {
    const secureRel = getSecureRelForTarget({ href, target, rel });

    return (
      <a
        tabIndex={isAriaHidden ? -1 : undefined}
        className={classes}
        href={href}
        target={target}
        rel={secureRel}
        ref={buttonRef as Ref<HTMLAnchorElement>}
        {...(rest as AnchorHTMLAttributes<HTMLAnchorElement>)}>
        {buttonIcon}
      </a>
    );
  }

  let buttonType: ButtonHTMLAttributes<HTMLButtonElement>['type'];
  return (
    <button
      tabIndex={isAriaHidden ? -1 : undefined}
      disabled={isDisabled}
      className={classes}
      type={type as typeof buttonType}
      ref={buttonRef as Ref<HTMLButtonElement>}
      {...(rest as ButtonHTMLAttributes<HTMLButtonElement>)}>
      {buttonIcon}
    </button>
  );
};
