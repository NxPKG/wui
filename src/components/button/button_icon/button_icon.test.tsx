/*
 * Copyright 2022 Wazuh Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * NOTICE: THIS FILE HAS BEEN MODIFIED BY WAZUH INC UNDER COMPLIANCE WITH THE APACHE 2.0 LICENSE FROM THE ORIGINAL WORK
 * OF THE COMPANY Elasticsearch B.V.
 *
 * THE FOLLOWING IS THE COPYRIGHT OF THE ORIGINAL DOCUMENT:
 *
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { render, mount } from 'enzyme';
import { requiredProps } from '../../../test/required_props';

import { WuiButtonIcon, COLORS } from './button_icon';

describe('WuiButtonIcon', () => {
  test('is rendered', () => {
    const component = render(<WuiButtonIcon {...requiredProps} />);

    expect(component).toMatchSnapshot();
  });

  describe('props', () => {
    describe('isDisabled', () => {
      it('is rendered', () => {
        const component = render(
          <WuiButtonIcon aria-label="button" isDisabled />
        );

        expect(component).toMatchSnapshot();
      });

      it('renders a button even when href is defined', () => {
        const component = render(
          <WuiButtonIcon aria-label="button" href="#" isDisabled />
        );

        expect(component).toMatchSnapshot();
      });
    });

    describe('iconType', () => {
      it('is rendered', () => {
        const component = render(
          <WuiButtonIcon aria-label="button" iconType="user" />
        );

        expect(component).toMatchSnapshot();
      });
    });

    describe('color', () => {
      COLORS.forEach(color => {
        test(`${color} is rendered`, () => {
          const component = render(
            <WuiButtonIcon aria-label="button" color={color} />
          );

          expect(component).toMatchSnapshot();
        });
      });
    });

    describe('href', () => {
      it('secures the rel attribute when the target is _blank', () => {
        const component = render(
          <WuiButtonIcon aria-label="button" href="#" target="_blank" />
        );

        expect(component).toMatchSnapshot();
      });
    });

    describe('onClick', () => {
      it('supports onClick and href', () => {
        const handler = jest.fn();
        const component = mount(
          <WuiButtonIcon aria-label="hoi" href="#" onClick={handler} />
        );
        component.find('a').simulate('click');
        expect(handler.mock.calls.length).toEqual(1);
      });

      it('supports onClick as a button', () => {
        const handler = jest.fn();
        const component = mount(
          <WuiButtonIcon aria-label="hoi" onClick={handler} />
        );
        component.find('button').simulate('click');
        expect(handler.mock.calls.length).toEqual(1);
      });
    });
  });
});
