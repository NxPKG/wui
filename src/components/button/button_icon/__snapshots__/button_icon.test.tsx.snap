// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WuiButtonIcon is rendered 1`] = `
<button
  aria-label="aria-label"
  class="wuiButtonIcon wuiButtonIcon--primary testClass1 testClass2"
  data-test-subj="test subject string"
  type="button"
/>
`;

exports[`WuiButtonIcon props color accent is rendered 1`] = `
<button
  aria-label="button"
  class="wuiButtonIcon wuiButtonIcon--accent"
  type="button"
/>
`;

exports[`WuiButtonIcon props color danger is rendered 1`] = `
<button
  aria-label="button"
  class="wuiButtonIcon wuiButtonIcon--danger"
  type="button"
/>
`;

exports[`WuiButtonIcon props color disabled is rendered 1`] = `
<button
  aria-label="button"
  class="wuiButtonIcon wuiButtonIcon--disabled"
  type="button"
/>
`;

exports[`WuiButtonIcon props color ghost is rendered 1`] = `
<button
  aria-label="button"
  class="wuiButtonIcon wuiButtonIcon--ghost"
  type="button"
/>
`;

exports[`WuiButtonIcon props color primary is rendered 1`] = `
<button
  aria-label="button"
  class="wuiButtonIcon wuiButtonIcon--primary"
  type="button"
/>
`;

exports[`WuiButtonIcon props color subdued is rendered 1`] = `
<button
  aria-label="button"
  class="wuiButtonIcon wuiButtonIcon--subdued"
  type="button"
/>
`;

exports[`WuiButtonIcon props color success is rendered 1`] = `
<button
  aria-label="button"
  class="wuiButtonIcon wuiButtonIcon--success"
  type="button"
/>
`;

exports[`WuiButtonIcon props color text is rendered 1`] = `
<button
  aria-label="button"
  class="wuiButtonIcon wuiButtonIcon--text"
  type="button"
/>
`;

exports[`WuiButtonIcon props color warning is rendered 1`] = `
<button
  aria-label="button"
  class="wuiButtonIcon wuiButtonIcon--warning"
  type="button"
/>
`;

exports[`WuiButtonIcon props href secures the rel attribute when the target is _blank 1`] = `
<a
  aria-label="button"
  class="wuiButtonIcon wuiButtonIcon--primary"
  href="#"
  rel="noopener noreferrer"
  target="_blank"
/>
`;

exports[`WuiButtonIcon props iconType is rendered 1`] = `
<button
  aria-label="button"
  class="wuiButtonIcon wuiButtonIcon--primary"
  type="button"
>
  <div
    aria-hidden="true"
    class="wuiButtonIcon__icon"
    data-wuiicon-type="user"
  />
</button>
`;

exports[`WuiButtonIcon props isDisabled is rendered 1`] = `
<button
  aria-label="button"
  class="wuiButtonIcon wuiButtonIcon--primary"
  disabled=""
  type="button"
/>
`;

exports[`WuiButtonIcon props isDisabled renders a button even when href is defined 1`] = `
<button
  aria-label="button"
  class="wuiButtonIcon wuiButtonIcon--primary"
  disabled=""
  type="button"
/>
`;
